parsing/dialects/dialect1/parserImpls.ftl

```diff
@@ -2382,3 +2382,164 @@ SqlCreateJoinIndex SqlCreateJoinIndex() :
             select, indices);
     }
 }
+
+SqlCall CaseN() :
+{
+    SqlNodeList nodes = new SqlNodeList(getPos());
+    SqlNode e;
+    NoCaseUnknown extraPartitionOption = null;
+}
+{
+    <CASE_N>
+    <LPAREN>
+    e = Expression(ExprContext.ACCEPT_SUB_QUERY)
+    {
+        nodes.add(e);
+    }
+    (
+        <COMMA>
+        e = Expression(ExprContext.ACCEPT_SUB_QUERY)
+        {
+            nodes.add(e);
+        }
+    )*
+    [
+        <COMMA>
+        (
+            LOOKAHEAD(3)
+            <NO> <CASE> <OR> <UNKNOWN>
+            { extraPartitionOption = NoCaseUnknown.NO_CASE_OR_UNKNOWN; }
+        |
+            LOOKAHEAD(3)
+            <NO> <CASE> <COMMA> <UNKNOWN>
+            { extraPartitionOption = NoCaseUnknown.NO_CASE_COMMA_UNKNOWN; }
+        |
+            <NO> <CASE> { extraPartitionOption = NoCaseUnknown.NO_CASE; }
+        |
+            <UNKNOWN> { extraPartitionOption = NoCaseUnknown.UNKNOWN; }
+        )
+    ]
+    <RPAREN>
+    {
+        return new SqlCaseN(getPos(), nodes, extraPartitionOption);
+    }
+}
+
+SqlCall RangeN() :
+{
+    SqlNode testExpression;
+    SqlNode range;
+    SqlNode startLiteral = null;
+    boolean startAsterisk = false;
+    SqlNode endLiteral = null;
+    boolean endAsterisk = false;
+    SqlNode eachSizeLiteral = null;
+    SqlNodeList rangeList = new SqlNodeList(getPos());
+    NoRangeUnknown extraPartitionOption = null;
+}
+{
+    <RANGE_N>
+    <LPAREN>
+    testExpression = CompoundIdentifier()
+    <BETWEEN>
+    (
+        <STAR> { startAsterisk = true; }
+        [
+            <AND>
+            (
+                <STAR> { endAsterisk = true; }
+            |
+                endLiteral = Literal()
+            )
+        ]
+    |
+        startLiteral = Literal()
+        [
+            <AND>
+            (
+                <STAR> { endAsterisk = true; }
+            |
+                endLiteral = Literal()
+            )
+        ]
+        [
+            <EACH>
+            eachSizeLiteral = Literal()
+        ]
+    )
+
+    { rangeList.add(new SqlRangeNStartEnd(getPos(), startLiteral, endLiteral
+          , eachSizeLiteral, startAsterisk, endAsterisk));
+          startLiteral = null;
```

Comment:
nit: Do we need to explicitly set these when they're already being set to these values in the initialization list?

Reply:
The current implementation require these variable to be reset in the middle of the parsing logic, otherwise it would have side effect on the next parsing of RangeNStartEnd. The RangeNStartEnd is now parsed inside RangeN. It was originally made to be its own function. However, since we need the parser to enforce certain constraints "only leading range can have asterisk as start" and "any range that have asterisk as start cannot have each clause", I move back the logic to RangeN() for easier modification.

A way to reduce this complexity as I see it now would be extract the non-first range to a separated function. Then we don't need to reset all these variable for every new range appears.

Reply:
Separating it out into a helper function to reduce complexity sounds like a good idea to me.

Reply:
Fixed in https://github.com/googleinterns/calcite/pull/212/commits/6b21e8112aa954786d22621cd70ecd1c18d4cde3